File: Nibbles/Snake.cs
namespace Nibbles
{
    using System;
    using System.Linq;
    using System.Collections.Generic;
    
    class Snake : Vector, ISnake
    {
        private static byte Count = 1;

        private int ticks = 0;

        private List<Vector> trail;

        public ISnakeBehavior Behavior { get; set; }
        
        public IReadOnlyCollection<IVector> Trail 
        { 
            get
            {
                return trail.Select( vec => (IVector)vec ).ToList().AsReadOnly();
            }
        }

        public int Ticks
        { 
            get{ return ticks; } 
        }

        public Snake()
        {
            Id = Count++;
            trail = new List<Vector>();
            Length = 1;
        }

        public int Id { get; set; }

        public int Length { get; set;}

        public void ChangeDirection(Direction direction)
        {
            if(Direction != direction || Trail.Count == 0)
            {
                trail.Add(new Vector{X = X, Y = Y, Direction = direction});
                Direction = direction;
            }
        }

        public Vector ReduceLength(int count)
        {
            Length = Length - count;
            Vector first = trail.First();
            Vector second = Trail.Count > 1 ? trail[1] : null;
            var newFirst = first.MoveNew(first.Direction);
            first.SetPosition(newFirst);
            if(second != null && 
               first.X == second.X && first.Y == second.Y)
            {
                trail.Remove(first);
            }
            return first;
        }

        public void Step()
        {
            ticks++;
            Length++;
            Move();
        }

        public Vector GetTailPosition()
        {
            Vector first = trail.First();
            return new Vector{X=first.X, Y=first.Y, Direction=first.Direction};
        }

        public bool WillHitNextStep(Space space)
        {
            IPosition pos = MoveNew(Direction);
            if(pos.IsValid(space.TopX, space.TopY)
               && (space[pos.X, pos.Y] == 0))
            {
                return false;
            }
            return true;
        }

        public int WillHitNextStepInfo(Space space)
        {
            IPosition pos = MoveNew(Direction);
            if(!pos.IsValid(space.TopX, space.TopY))
            {
                return -1;
            }
            return space[pos.X, pos.Y];
        }

        public void PrintSnake()
        {
            Console.WriteLine($"  {Id}:({Ticks}): x:{X}, y:{Y}, {Direction}, l:{Length}");
            foreach(var vec in Trail)
            {
                Console.WriteLine($"\t{vec.X}, {vec.Y} {vec.Direction}");
            }
        }
    }
}

File: wnibbles/src/client/Snake.ts
import {Vector} from './Vector';
import {Space} from './Space';
import {ISnakeBehavior} from './ISnakeBehavior';
import {Color} from './Nibbles';
import {Direction} from './Direction';

export class Snake extends Vector {
    id: number;
    ticks: number;
    trail: Array<Vector>;
    length: number;
    color: Color;
    behavior: ISnakeBehavior;

    constructor() {
        super();
        this.trail = new Array<Vector>();
    }

    reduceLength(count: number): Vector {
        this.length -= count;

        let first = this.trail[0];
        let second = this.trail.length > 1 ? this.trail[1] : null;
        let newFirst = first.moveNewDirection(first.direction);
        first.x = newFirst.x;
        first.y = newFirst.y;
        if(second != null && 
            first.x == second.x && first.y == second.y) {
            this.trail.splice(0,1);
        }
        return first;
    }

    getTailPosition(): Vector {
        let first = this.trail[0];
        let newVec = new Vector();
        newVec.x = first.x;
        newVec.y = first.y;
        newVec.direction = first.direction;
        return newVec;
    }

    step() {
        this.ticks++;
        this.length++;
        this.move();
    }

    willHitNextStep(space:Space):boolean {
        let pos = this.moveNewDirection(this.direction);
        if(pos.isValidInBounds(space.topX, space.topY)
            && (space.map[pos.x][pos.y] == 0)) {
            return false;
        }
        return true;
    }

    willHitNextStepInfo(space: Space) : number {
        let pos = this.moveNewDirection(this.direction);
        if(!pos.isValidInBounds(space.topX, space.topY)) {
            return -1;
        }
        return space.map[pos.x][pos.y];
    }

    changeDirection(direction: Direction) {
        if(this.direction != direction || this.trail.length == 0) {
            let vec = new Vector();
            vec.x = this.x;
            vec.y = this.y;
            vec.direction = direction;
            this.trail.push(vec);
            this.direction = direction;
        }
    }

}
